service/history/workflow/mutable_state_impl.go

```diff
@@ -938,30 +943,63 @@ func (ms *MutableStateImpl) GetStartVersion() (int64, error) {
 		if err != nil {
 			return 0, err
 		}
-		firstItem, err := versionhistory.GetFirstVersionHistoryItem(versionHistory)
-		if err != nil {
-			return 0, err
+
+		if !versionhistory.IsEmptyVersionHistory(versionHistory) {
+			firstItem, err := versionhistory.GetFirstVersionHistoryItem(versionHistory)
+			if err != nil {
+				return 0, err
+			}
+			return firstItem.GetVersion(), nil
 		}
-		return firstItem.GetVersion(), nil
+	}
+
+	// We can't check TransitionHistory before VersionHistories because old workflows (mutable state) may
+	// have transition history enabled while they are running, so the first item in the transition history
+	// is not the actual start version.
+	if ms.transitionHistoryEnabled && len(ms.executionInfo.TransitionHistory) != 0 {
+		return ms.executionInfo.TransitionHistory[0].GetNamespaceFailoverVersion(), nil
 	}
 
 	return common.EmptyVersion, nil
 }
 
 func (ms *MutableStateImpl) GetCloseVersion() (int64, error) {
-	if ms.IsWorkflowExecutionRunning() {
+	// Do NOT use ms.IsWorkflowExecutionRunning() for the check.
+	// Zombie workflow is not considered running but also not closed.
+	if ms.executionState.State != enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED {
 		return common.EmptyVersion, serviceerror.NewInternal("GetCloseVersion: workflow is still running")
 	}
 
-	// here we assume that closing a workflow must generate an event
-
 	// if workflow is closing in the current transation,
 	// then the last event is closed event and the event version is the close version
 	if lastEventVersion, ok := ms.hBuilder.LastEventVersion(); ok {
 		return lastEventVersion, nil
 	}
 
-	return ms.GetLastEventVersion()
+	if ms.executionInfo.VersionHistories != nil {
+		isEmpty, err := versionhistory.IsCurrentVersionHistoryEmpty(ms.executionInfo.VersionHistories)
+		if err != nil {
+			return common.EmptyVersion, err
+		}
+		if !isEmpty {
+			return ms.GetLastEventVersion()
+		}
+	}
+
+	if ms.transitionHistoryEnabled {
+		if ms.executionStateUpdated {
+			// closing in the current transaction
+			return ms.GetCurrentVersion(), nil
+		}
+
+		// once the mutable state is closed, the execution state will no longer be updated.
+		// so the last update version is the close version.
+		if ms.executionState.LastUpdateVersionedTransition != nil {
+			return ms.executionState.LastUpdateVersionedTransition.NamespaceFailoverVersion, nil
+		}
+	}
+
+	return common.EmptyVersion, nil
```

Comment:
is this expected in some case to return empty version?

Reply:
I feel we have bunch of logic probably trying to handle backward compatibility (by checking for version history is nil) and return EmptyVersion. Like GetCurrentVersion(). If you have any context on that, plz let me know.

I'd say for any new workflows/chasm runs, it's not expected to reach this line (EmptyVersion can still be returned above for local namespace). For very old workflows, i am not sure.